src/main/java/com/fivetran/external/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ParentsFirstShardPrioritization.java

```diff
@@ -48,38 +48,38 @@ public ParentsFirstShardPrioritization(int maxDepth) {
     }
 
     @Override
-    public List<ShardInfo> prioritize(List<ShardInfo> original) {
-        Map<String, ShardInfo> shards = new HashMap<>();
-        for (ShardInfo shardInfo : original) {
+    public <T extends PriorizeableShard> List<T> prioritize(List<T> original) {
+        Map<String, T> shards = new HashMap<>();
+        for (T shardInfo : original) {
             shards.put(shardInfo.getShardId(),
                     shardInfo);
         }
 
-        Map<String, SortingNode> processedNodes = new HashMap<>();
+        Map<String, SortingNode<T>> processedNodes = new HashMap<>();
 
-        for (ShardInfo shardInfo : original) {
+        for (T shardInfo : original) {
             populateDepth(shardInfo.getShardId(),
                     shards,
                     processedNodes);
         }
 
-        List<ShardInfo> orderedInfos = new ArrayList<>(original.size());
+        List<T> orderedInfos = new ArrayList<>(original.size());
 
         List<SortingNode> orderedNodes = new ArrayList<>(processedNodes.values());
         Collections.sort(orderedNodes);
 
         for (SortingNode sortingTreeNode : orderedNodes) {
             // don't process shards with depth > maxDepth
             if (sortingTreeNode.getDepth() <= maxDepth) {
-                orderedInfos.add(sortingTreeNode.shardInfo);
+                orderedInfos.add((T) sortingTreeNode.shardInfo);
             }
         }
         return orderedInfos;
     }
 
-    private int populateDepth(String shardId,
-            Map<String, ShardInfo> shards,
-            Map<String, SortingNode> processedNodes) {
+    private <T extends PriorizeableShard> int populateDepth(String shardId,
```

Comment:
Note this method is recursive. I tested it out and it handled a lineage of 4500 shards.

Reply:
If you are concerned that we may hit the recursion limit (seems unlikely) we could do this iteratively instead.